# ETL (Extract, Transform, Load)

This dataset contains detailed information on over 125,000 restaurants across 31 major cities in Europe, collected from Tripadvisor. Each entry includes:

- City location

- Cuisine (single or multiple cuisines per restaurant)

- Rating (on a 1–5 scale)

- Ranking within the city

- Price Range (Low, Mid, High)

- Number of Reviews

- Sample of customer reviews and their dates

- Tripadvisor URL and unique restaurant ID

The dataset supports a wide range of analyses, including cuisine trends, pricing patterns, customer preferences, and regional comparisons. While rich in content, it also contains missing values that require preprocessing for accurate analysis.

## Objectives
The objective of this step is to extract, clean, and transform raw data into a structured format suitable for analysis and visualization, ensuring consistency, accuracy, and usability throughout the project.

## Inputs
The dataset was obtained from [Kaggle](https://www.kaggle.com/datasets/damienbeneschi/krakow-ta-restaurans-data-raw/data)

## Outputs
The cleaned csv file found [here]()

# ETL Process

- Load the dataset
- Understand dataset structure and content
- Convert data types
- Clean the dataset
- Add country column
- Add cuisine counts column
- Save the clean dataset as a csv file

---

# Change working directory
Change the working directory from its current folder to its parent folder as the notebooks will be stored in a subfolder
* We access the current directory with os.getcwd()

In [1]:
import os
current_dir = os.getcwd()
current_dir

'c:\\Users\\amron\\Desktop\\euro-dine-insights\\jupyter_notebooks'

Make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [2]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [3]:
current_dir = os.getcwd()
current_dir

'c:\\Users\\amron\\Desktop\\euro-dine-insights'

Changing path directory to the dataset

In [4]:
raw_data_dir = os.path.join(current_dir, 'data_set/raw') #path directory

processed_data_dir = os.path.join(current_dir, 'data_set/processed') #path directory


---

# Import packages

In [5]:
import numpy as np #import numpy
import pandas as pd #import pandas
import matplotlib.pyplot as plt #import matplotlib
import seaborn as sns #import seaborn
import plotly.express as px # import plotly
sns.set_style('whitegrid') #set style for visuals

---

# Load the raw dataset

In [6]:
#load the dataset
import pandas as pd
df = pd.read_csv(os.path.join(raw_data_dir, 'TA_restaurants_curated.csv'))

The raw dataset is loaded using Pandas for ETL process

---

# Understand the dataset structure and content

In [7]:
#displaying data
df.head() 

Unnamed: 0                        Name       City  \
0           0  Martine of Martine's Table  Amsterdam   
1           1         De Silveren Spiegel  Amsterdam   
2           2                     La Rive  Amsterdam   
3           3                    Vinkeles  Amsterdam   
4           4   Librije's Zusje Amsterdam  Amsterdam   

                                       Cuisine Style  Ranking  Rating  \
0                    ['French', 'Dutch', 'European']      1.0     5.0   
1  ['Dutch', 'European', 'Vegetarian Friendly', '...      2.0     4.5   
2  ['Mediterranean', 'French', 'International', '...      3.0     4.5   
3  ['French', 'European', 'International', 'Conte...      4.0     5.0   
4  ['Dutch', 'European', 'International', 'Vegeta...      5.0     4.5   

  Price Range  Number of Reviews  \
0    $$ - $$$              136.0   
1        $$$$              812.0   
2        $$$$              567.0   
3        $$$$              564.0   
4        $$$$              316.0   

                                             Reviews  \
0  [['Just like home', 'A Warm Welcome to Wintry ...   
1  [['Great food and staff', 'just perfect'], ['0...   
2  [['Satisfaction', 'Delicious old school restau...   
3  [['True five star dinner', 'A superb evening o...   
4  [['Best meal.... EVER', 'super food experience...   

                                              URL_TA      ID_TA  
0  /Restaurant_Review-g188590-d11752080-Reviews-M...  d11752080  
1  /Restaurant_Review-g188590-d693419-Reviews-De_...    d693419  
2  /Restaurant_Review-g188590-d696959-Reviews-La_...    d696959  
3  /Restaurant_Review-g188590-d1239229-Reviews-Vi...   d1239229  
4  /Restaurant_Review-g188590-d6864170-Reviews-Li...   d6864170

Upon loading the dataset, we observed an Unnamed: 0 column, which is a duplicate of the default index. Since it does not carry any additional information, it will be dropped in the data cleaning step to avoid redundancy.

In [8]:
#check data info
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 125527 entries, 0 to 125526
Data columns (total 11 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   Unnamed: 0         125527 non-null  int64  
 1   Name               125527 non-null  object 
 2   City               125527 non-null  object 
 3   Cuisine Style      94176 non-null   object 
 4   Ranking            115876 non-null  float64
 5   Rating             115897 non-null  float64
 6   Price Range        77672 non-null   object 
 7   Number of Reviews  108183 non-null  float64
 8   Reviews            115911 non-null  object 
 9   URL_TA             125527 non-null  object 
 10  ID_TA              125527 non-null  object 
dtypes: float64(3), int64(1), object(7)
memory usage: 10.5+ MB


The raw dataset contains 125,527 rows and 11 columns. It has a significant number of missing values in several key columns such as Cuisine Style, Price Range, and Number of Reviews. Additionally, some data types may need to be converted to prepare the data for analysis.

---

# Clean the data

### 1. Dropping Unnamed: 0 column

This column provides no useful information and can be removed to clean up the dataset.

In [9]:
#dropping Unnamed: 0 column
df = df.drop(columns=['Unnamed: 0'])

In [10]:
#display data after Unnamed: 0 column dropped
df.head(1)

Name       City                    Cuisine Style  \
0  Martine of Martine's Table  Amsterdam  ['French', 'Dutch', 'European']   

   Ranking  Rating Price Range  Number of Reviews  \
0      1.0     5.0    $$ - $$$              136.0   

                                             Reviews  \
0  [['Just like home', 'A Warm Welcome to Wintry ...   

                                              URL_TA      ID_TA  
0  /Restaurant_Review-g188590-d11752080-Reviews-M...  d11752080

### 2. Identifying and handling missing values

In [11]:
#check for null values
df.isnull().sum() 

Name                     0
City                     0
Cuisine Style        31351
Ranking               9651
Rating                9630
Price Range          47855
Number of Reviews    17344
Reviews               9616
URL_TA                   0
ID_TA                    0
dtype: int64

The dataset contains missing values in several columns:

- Cuisine Style: 31,351 nulls

- Ranking: 9,651 nulls

- Rating: 9,630 nulls

- Price Range: 47,855 nulls

- Number of Reviews: 17,344 nulls

- Reviews: 9,616 nulls

To ensure reliable analysis, these missing values should be addressed through data cleaning techniques like removal or filling with default values.

In [12]:
#dropping rows with null values
df = df.dropna()

In [13]:
#check null values again
df.isnull().sum()

Name                 0
City                 0
Cuisine Style        0
Ranking              0
Rating               0
Price Range          0
Number of Reviews    0
Reviews              0
URL_TA               0
ID_TA                0
dtype: int64

All rows containing null values have been removed from the dataset to ensure data completeness and improve the quality of analysis.

### 3. Identifying duplicates in the dataset

In [14]:
#check for duplicates
df.duplicated(keep=False).sum() 

8

The result showed 8 duplicate entries, indicating that some rows contain identical data and need to be removed to avoid redundancy and ensure accurate analysis.

In [15]:
#display the 8 duplicate rows for checking
df[df.duplicated(keep=False)].head(8)

Name    City  \
122677               Rolls Rolls  Warsaw   
122678               Rolls Rolls  Warsaw   
123217                  Flow Bar  Warsaw   
123218                  Flow Bar  Warsaw   
123307              Basico Pizza  Warsaw   
123308              Basico Pizza  Warsaw   
123487  Restauracja Olive Garden  Warsaw   
123488  Restauracja Olive Garden  Warsaw   

                                            Cuisine Style  Ranking  Rating  \
122677            ['Sushi', 'Asian', 'Fusion', 'Healthy']   1171.0     4.0   
122678            ['Sushi', 'Asian', 'Fusion', 'Healthy']   1171.0     4.0   
123217                      ['Bar', 'Polish', 'European']   1711.0     3.0   
123218                      ['Bar', 'Polish', 'European']   1711.0     3.0   
123307                                        ['Italian']   1801.0     3.5   
123308                                        ['Italian']   1801.0     3.5   
123487  ['Mediterranean', 'Middle Eastern', 'Italian',...   1981.0     4.0   
123488  ['Mediterranean', 'Middle Eastern', 'Italian',...   1981.0     4.0   

       Price Range  Number of Reviews  \
122677           $               14.0   
122678           $               14.0   
123217    $$ - $$$               73.0   
123218    $$ - $$$               73.0   
123307           $                3.0   
123308           $                3.0   
123487    $$ - $$$               16.0   
123488    $$ - $$$               16.0   

                                                  Reviews  \
122677  [['Great Rolls Sushi', 'Overpiced, overhyped, ...   
122678  [['Great Rolls Sushi', 'Overpiced, overhyped, ...   
123217  [['Nice spot', 'Good Position'], ['09/03/2017'...   
123218  [['Nice spot', 'Good Position'], ['09/03/2017'...   
123307                                           [[], []]   
123308                                           [[], []]   
123487  [['Good restaurant for one meal.', 'Olive Gard...   
123488  [['Good restaurant for one meal.', 'Olive Gard...   

                                                   URL_TA      ID_TA  
122677  /Restaurant_Review-g274856-d9756783-Reviews-Ro...   d9756783  
122678  /Restaurant_Review-g274856-d9756783-Reviews-Ro...   d9756783  
123217  /Restaurant_Review-g274856-d2351903-Reviews-Fl...   d2351903  
123218  /Restaurant_Review-g274856-d2351903-Reviews-Fl...   d2351903  
123307  /Restaurant_Review-g274856-d10316215-Reviews-B...  d10316215  
123308  /Restaurant_Review-g274856-d10316215-Reviews-B...  d10316215  
123487  /Restaurant_Review-g274856-d3396817-Reviews-Re...   d3396817  
123488  /Restaurant_Review-g274856-d3396817-Reviews-Re...   d3396817

In [16]:
#drop duplicates
df = df.drop_duplicates()

In [17]:
#check for duplicates again to make sure the duplicated rows are dropped
df.duplicated(keep=False).sum() 

0

 ### 4. Adjusting columns names and remove spacing

In [18]:
#renaming columns
df.rename(columns={
    'Cuisine Style': 'Cuisine',
    'Number of Reviews': 'Number_of_Reviews',
    'Price Range': 'Price_Range'}, inplace=True) 

In [19]:
#check adjusted columns names
df.columns

Index(['Name', 'City', 'Cuisine', 'Ranking', 'Rating', 'Price_Range',
       'Number_of_Reviews', 'Reviews', 'URL_TA', 'ID_TA'],
      dtype='object')

Selected column names were renamed for consistency and improved readability.